hal/halx86/apic/apic.c

```diff
@@ -671,87 +858,470 @@ HalDisableSystemInterrupt(
+            break;
+
+        pPrcbVector[0] = Idx - 1;
+        KeSetCurrentIrql(Irql - 1);
+
+        HalpGenerateInterrupt(Vector);
+        //HalpTotalReplayed++;
+    }
+
+    KeSetCurrentIrql(NewIrql);
+
+    if (EFlags & EFLAGS_INTERRUPT_MASK)
+        _enable();
 }
 
+VOID
+NTAPI
+HalpDispatchSoftwareInterrupt(_In_ KIRQL Irql,
+                              _In_ PKTRAP_FRAME TrapFrame)
+{
+    PHALP_PCR_HAL_RESERVED HalReserved;
+    ULONG EFlags;
+
+    EFlags = __readeflags();
+    KeGetPcr()->Irql = Irql;
+
+    if (!(EFlags & EFLAGS_INTERRUPT_MASK))
+        _enable();
+
+    HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+    if (Irql == APC_LEVEL)
+    {
+        HalReserved->ApcRequested = FALSE;
+        KiDeliverApc(0, 0, (PKTRAP_FRAME)TrapFrame);
+    }
+    else if (Irql == DISPATCH_LEVEL)
+    {
+        HalReserved->DpcRequested = FALSE;
+        KiDispatchInterrupt();
+    }
+    else
+    {
+        DbgBreakPoint();
+    }
+
+    if (!(EFlags & EFLAGS_INTERRUPT_MASK))
+        _disable();
+}
+
+VOID
+FASTCALL
+HalpCheckForSoftwareInterrupt(_In_ KIRQL NewIrql,
+                              _In_ PKTRAP_FRAME TrapFrame)
+{
+    PHALP_PCR_HAL_RESERVED HalReserved;
+    BOOLEAN ApcRequested;
+    BOOLEAN DpcRequested;
+
+    HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+
+    ApcRequested = HalReserved->ApcRequested;
+    DpcRequested = HalReserved->DpcRequested;
+
+    if (NewIrql)
+    {
+        if (NewIrql == APC_LEVEL && HalReserved->DpcRequested)
+        {
+            do
+            {
+                HalpDispatchSoftwareInterrupt(DISPATCH_LEVEL, TrapFrame);
+                KeSetCurrentIrql(APC_LEVEL);
+                HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+            }
+            while (HalReserved->DpcRequested);
+        }
+
+        return;
+    }
+
+    while (ApcRequested | DpcRequested)
+    {
+        if (DpcRequested)
+        {
+            HalpDispatchSoftwareInterrupt(DISPATCH_LEVEL, TrapFrame);
+        }
+        else
+        {
+            HalpDispatchSoftwareInterrupt(APC_LEVEL, TrapFrame);
+        }
+
+        KeSetCurrentIrql(PASSIVE_LEVEL);
+
+        HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+        ApcRequested = HalReserved->ApcRequested;
+        DpcRequested = HalReserved->DpcRequested;
+    }
+}
+
+BOOLEAN
+NTAPI
+HalBeginSystemInterrupt(_In_ KIRQL NewIrql,
+                        _In_ ULONG SystemVector,
+                        _Out_ PKIRQL OutOldIrql)
+{
+    PUCHAR pPrcbVector;
+    UCHAR Idx;
+    KIRQL OldIrql;
+
+    OldIrql = KeGetCurrentIrql();
+
+    if (OldIrql < HalpVectorToIRQL[(UCHAR)SystemVector >> 4])
+    {
+        *OutOldIrql = OldIrql;
+        KeSetCurrentIrql(NewIrql);
+
+        _enable();
+        return TRUE;
+    }
+
+    pPrcbVector = (PUCHAR)KeGetCurrentPrcb()->HalReserved;
+
+    Idx = pPrcbVector[0];
+    pPrcbVector[Idx + 1] = SystemVector;
+    pPrcbVector[0] = Idx + 1;
```

Comment:
You seem to use `HalReserved` field of the PKRCB. What is the meaning of the first byte and the next `UINT32`s?

Reply:
If the IRQL (which corresponds to the incoming (new) SystemVector) is less than the current IRQL, then this vector is stored in the "interrupt stack" (in HalReserved). This will subsequently be fetched (in a loop) and executed by HalpGenerateInterrupt (Vector) (asm - "int Vector").
By the way, don't confuse HalReserved in Pcr and Prcb.

first byte - current position in "interrupt stack"
next - stack values (SystemVector`s)